src/control/server/instance_drpc.go

```diff
@@ -215,86 +216,88 @@ func (ei *EngineInstance) getSmdDetails(smd *ctlpb.SmdDevResp_Device) (*storage.
-	}
-
+func updateCtrlrHealth(pbStats *ctlpb.BioHealthResp, ctrlr *storage.NvmeController) error {
 	ctrlr.HealthStats = new(storage.NvmeHealth)
 	if err := convert.Types(pbStats, ctrlr.HealthStats); err != nil {
-		return false, errors.Wrap(err, "convert health stats")
+		return errors.Wrap(err, "convert health stats")
 	}
 
-	ei.log.Debugf("%s: health stats updated", msg)
-	return true, nil
+	return nil
 }
 
 // updateInUseBdevs updates-in-place the input list of controllers with new NVMe health stats and
 // SMD metadata info.
 //
 // Query each SmdDevice on each I/O Engine instance for health stats and update existing controller
 // data in ctrlrMap using PCI address key.
-func (ei *EngineInstance) updateInUseBdevs(ctx context.Context, ctrlrMap map[string]*storage.NvmeController) error {
+func (ei *EngineInstance) updateInUseBdevs(ctx context.Context, ctrlrMap map[string]*storage.NvmeController) (err error) {
+	defer func() {
+		err = errors.Wrapf(err, "instance %d", ei.Index())
+	}()
+
+	// Clear SMD info for controllers in ctrlrMap, populate smd info from scratch.
+	for _, ctrlr := range ctrlrMap {
+		ctrlr.SmdDevices = []*storage.SmdDevice{}
+	}
+
 	smdDevs, err := ei.ListSmdDevices(ctx, new(ctlpb.SmdDevReq))
 	if err != nil {
-		return errors.Wrapf(err, "instance %d listSmdDevices()", ei.Index())
+		return errors.Wrapf(err, "list smd devices")
 	}
+	ei.log.Debugf("engine %d: smdDevs %+v", ei.Index(), smdDevs)
 
 	hasUpdatedHealth := make(map[string]bool)
 	for _, smd := range smdDevs.Devices {
+		msg := fmt.Sprintf("instance %d: smd %s: ctrlr %s", ei.Index(), smd.Uuid,
+			smd.TrAddr)
+
 		ctrlr, exists := ctrlrMap[smd.GetTrAddr()]
 		if !exists {
-			return errors.Errorf("instance %d: smd %s: unknown controller %s",
-				ei.Index(), smd.GetUuid(), smd.GetTrAddr())
+			ei.log.Errorf("%s: ctrlr not found", msg)
+			continue
 		}
 
 		smdDev, err := ei.getSmdDetails(smd)
 		if err != nil {
-			return errors.Wrapf(err, "collect smd info for ctrlr %s", ctrlr.PciAddr)
+			return errors.Wrapf(err, "%s: collect smd info", msg)
 		}
 
-		// multiple updates for the same key expected when more than one controller
-		// namespaces (and resident blobstores) exist, stats will be the same for each
-		// so only pass valid ctrlr reference when stats haven't yet been updated
-		var ctrlrRef *storage.NvmeController
-		if _, already := hasUpdatedHealth[ctrlr.PciAddr]; !already {
-			ctrlrRef = ctrlr
+		pbStats, err := ei.GetBioHealth(ctx, &ctlpb.BioHealthReq{DevUuid: smdDev.UUID})
+		if err != nil {
+			// Only log error if error indicates non-existent health and the SMD entity
+			// has abnormal state.
+			status, ok := errors.Cause(err).(drpc.DaosStatus)
+			if ok && status == drpc.DaosNonexistant && !smdDev.NvmeState.IsNormal() {
+				ei.log.Debugf("%s: stats not found (device state: %q), skip update",
+					msg, smdDev.NvmeState.String())
+			} else {
+				ei.log.Errorf("%s: fetch stats: %s", msg, err.Error())
+			}
+			ctrlr.UpdateSmd(smdDev)
```

Comment:
From what I understand, smdDev information such as AvailBytes will be set to their default value without error returned. Could it be an issue, for functions using these information afterwards ?

Reply:
this is a best effort, there are some situations where stats will not be available (for example where health statistics cannot be returned for example when a new device has been inserted but is not yet being used by the engine (SMD "NEW") state. In these cases we want to update SmdInfo even with unpopulated (zeored) stats because we don't have valid information for usage but we still want static details like UUID.

Reply:
OK, thanks for the explanation.